// @medusajs/core-flows/dist/cart/workflows/get-variants-and-items-with-prices.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getVariantsAndItemsWithPrices = exports.getVariantsAndItemsWithPricesId = void 0;
const utils_1 = require("@medusajs/framework/utils");
const workflows_sdk_1 = require("@medusajs/framework/workflows-sdk");
const common_1 = require("../../common");
const steps_1 = require("../steps");
const fields_1 = require("../utils/fields");
const prepare_line_item_data_1 = require("../utils/prepare-line-item-data");
exports.getVariantsAndItemsWithPricesId = "get-variant-items-with-prices-workflow";
exports.getVariantsAndItemsWithPrices = (0, workflows_sdk_1.createWorkflow)(exports.getVariantsAndItemsWithPricesId, (input) => {
    const variantIds = (0, workflows_sdk_1.transform)({ cart: input.cart, items: input.items, variantIds: input.variants?.id }, (data) => {
        if (data.variantIds) {
            return data.variantIds;
        }
        return Array.from(new Set((data.cart.items ?? data.items ?? []).map((i) => i.variant_id))).filter((v) => !!v);
    });
    const cartPricingContext = (0, workflows_sdk_1.transform)({
        cart: input.cart,
        items: input.items,
        setPricingContextResult: input.setPricingContextResult,
    }, (data) => {
        const cart = data.cart;
        const baseContext = {
            ...(0, utils_1.filterObjectByKeys)(cart, fields_1.cartFieldsForPricingContext),
            customer: cart.customer,
            region: cart.region,
            ...(data.setPricingContextResult ? data.setPricingContextResult : {}),
            currency_code: cart.currency_code ?? cart.region?.currency_code,
            region_id: cart.region_id,
            customer_id: cart.customer_id,
        };
        return (data.items ?? cart.items ?? [])
            .filter((i) => i.variant_id)
            .map((item) => {
            const idLike = item.id ?? (0, utils_1.simpleHash)(JSON.stringify(item));
            return {
                id: idLike,
                variantId: item.variant_id,
                context: {
                    ...baseContext,
                    quantity: item.quantity,
                    is_custom_price: !!item.unit_price,
                },
            };
        });
    });
    const variantQueryFields = (0, workflows_sdk_1.transform)({ variants: input.variants }, (data) => {
        return data.variants?.fields ?? fields_1.productVariantsFields;
    });
    const { data: variantsData } = (0, common_1.useQueryGraphStep)({
        entity: "variants",
        fields: variantQueryFields,
        filters: {
            id: variantIds,
        },
        options: {
            cache: {
                enable: true,
            },
        },
    }).config({ name: "fetch-variants" });
    const calculatedPriceSets = (0, steps_1.getVariantPriceSetsStep)({
        data: cartPricingContext,
    });
    const variantsItemsWithPrices = (0, workflows_sdk_1.transform)({
        cart: input.cart,
        items: input.items,
        variantsData,
        calculatedPriceSets,
    }, ({ cart, items: inputItems, variantsData, calculatedPriceSets, }) => {
        const priceNotFound = [];
        const variantNotFoundOrPublished = [];
        const items = (inputItems ?? cart.items ?? []).map((item) => {
            const item_ = item;
            const idLike = item.id ?? (0, utils_1.simpleHash)(JSON.stringify(item));
            let calculatedPriceSet = calculatedPriceSets[idLike];
            if (!calculatedPriceSet) {
                calculatedPriceSet = calculatedPriceSets[item_.variant_id];
            }
            const isCustomPrice = item_.is_custom_price ?? (0, utils_1.isDefined)(item?.unit_price);
            if (!calculatedPriceSet && item_.variant_id && !isCustomPrice) {
                priceNotFound.push(item_.variant_id);
            }
            const variant = variantsData.find((v) => v.id === item.variant_id);
            if ((item.variant_id && !variant) || // variant specified but doesn't exist
                (variant &&
                    (!variant?.product?.status ||
                        variant.product.status !== utils_1.ProductStatus.PUBLISHED)) // variant exists but product is not published
            ) {
                variantNotFoundOrPublished.push(item_.variant_id);
            }
            if (variant) {
                variant.calculated_price = calculatedPriceSet;
            }
            const input = {
                item: item_,
                variant: variant,
                cartId: cart.id,
                unitPrice: item_.unit_price,
                isTaxInclusive: item_.is_tax_inclusive ??
                    calculatedPriceSet?.is_calculated_price_tax_inclusive,
                isCustomPrice: isCustomPrice,
            };
            if (variant && !isCustomPrice) {
                input.unitPrice = calculatedPriceSet.calculated_amount;
                input.isTaxInclusive =
                    calculatedPriceSet.is_calculated_price_tax_inclusive;
            }
            const preparedItem = (0, prepare_line_item_data_1.prepareLineItemData)(input);
            return {
                selector: { id: item_.id },
                data: preparedItem,
            };
        });
        if (variantNotFoundOrPublished.length > 0) {
            throw new utils_1.MedusaError(utils_1.MedusaError.Types.INVALID_DATA, `Variants ${variantNotFoundOrPublished.join(", ")} do not exist or belong to a product that is not published`);
        }
        if (priceNotFound.length > 0) {
            throw new utils_1.MedusaError(utils_1.MedusaError.Types.INVALID_DATA, `Variants with IDs ${priceNotFound.join(", ")} do not have a price`);
        }
        return { variants: variantsData, lineItems: items };
    });
    return new workflows_sdk_1.WorkflowResponse(variantsItemsWithPrices);
});
