// @medusajs/core-flows/dist/cart/workflows/refresh-cart-items.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.refreshCartItemsWorkflow = exports.refreshCartItemsWorkflowId = void 0;
const utils_1 = require("@medusajs/framework/utils");
const workflows_sdk_1 = require("@medusajs/framework/workflows-sdk");
const common_1 = require("../../common");
const locking_1 = require("../../locking");
const steps_1 = require("../steps");
const fields_1 = require("../utils/fields");
const schemas_1 = require("../utils/schemas");
const get_variants_and_items_with_prices_1 = require("./get-variants-and-items-with-prices");
const refresh_cart_shipping_methods_1 = require("./refresh-cart-shipping-methods");
const refresh_payment_collection_1 = require("./refresh-payment-collection");
const update_cart_promotions_1 = require("./update-cart-promotions");
const update_tax_lines_1 = require("./update-tax-lines");
const upsert_tax_lines_1 = require("./upsert-tax-lines");
exports.refreshCartItemsWorkflowId = "refresh-cart-items";
/**
 * This workflow refreshes a cart to ensure its prices, promotion codes, taxes, and other details are applied correctly. It's useful
 * after making a chnge to a cart, such as after adding an item to the cart or adding a promotion code.
 *
 * This workflow is used by other cart-related workflows, such as the {@link addToCartWorkflow} after an item
 * is added to the cart.
 *
 * You can use this workflow within your own customizations or custom workflows, allowing you to refresh the cart after making updates to it in your
 * custom flows.
 *
 * @example
 * const { result } = await refreshCartItemsWorkflow(container)
 * .run({
 *   input: {
 *     cart_id: "cart_123",
 *   }
 * })
 *
 * @summary
 *
 * Refresh a cart's details after an update.
 *
 * @property hooks.setPricingContext - This hook is executed before the cart is refreshed. You can consume this hook to return any custom context useful for the prices retrieval of the variants in the cart.
 *
 * For example, assuming you have the following custom pricing rule:
 *
 * ```json
 * {
 *   "attribute": "location_id",
 *   "operator": "eq",
 *   "value": "sloc_123",
 * }
 * ```
 *
 * You can consume the `setPricingContext` hook to add the `location_id` context to the prices calculation:
 *
 * ```ts
 * import { refreshCartItemsWorkflow } from "@medusajs/medusa/core-flows";
 * import { StepResponse } from "@medusajs/workflows-sdk";
 *
 * refreshCartItemsWorkflow.hooks.setPricingContext((
 *   { cart, items, additional_data }, { container }
 * ) => {
 *   return new StepResponse({
 *     location_id: "sloc_123", // Special price for in-store purchases
 *   });
 * });
 * ```
 *
 * The variants' prices will now be retrieved using the context you return.
 *
 * :::note
 *
 * Learn more about prices calculation context in the [Prices Calculation](https://docs.medusajs.com/resources/commerce-modules/pricing/price-calculation) documentation.
 *
 * :::
 *
 */
exports.refreshCartItemsWorkflow = (0, workflows_sdk_1.createWorkflow)({
    name: exports.refreshCartItemsWorkflowId,
    idempotent: false,
}, (input) => {
    (0, locking_1.acquireLockStep)({
        key: input.cart_id,
        timeout: 2,
        ttl: 10,
    });
    const setPricingContext = (0, workflows_sdk_1.createHook)("setPricingContext", {
        cart_id: input.cart_id,
        items: input.items,
        additional_data: input.additional_data,
    }, {
        resultValidator: schemas_1.pricingContextResult,
    });
    const setPricingContextResult = setPricingContext.getResult();
    (0, workflows_sdk_1.when)("force-refresh-calculate-prices", { input }, ({ input }) => {
        return !!input.force_refresh;
    }).then(() => {
        const { data: cart } = (0, common_1.useQueryGraphStep)({
            entity: "cart",
            fields: fields_1.cartFieldsForRefreshSteps,
            filters: { id: input.cart_id },
            pagination: {
                take: 1,
            },
            options: {
                isList: false,
            },
        });
        (0, steps_1.validateCartStep)({ cart });
        const { lineItems } = get_variants_and_items_with_prices_1.getVariantsAndItemsWithPrices.runAsStep({
            input: {
                cart,
                setPricingContextResult: setPricingContextResult,
            },
        });
        (0, steps_1.updateLineItemsStep)({
            id: cart.id,
            items: lineItems,
        });
    });
    const { data: refetchedCart } = (0, common_1.useQueryGraphStep)({
        entity: "cart",
        fields: fields_1.cartFieldsForRefreshSteps,
        filters: { id: input.cart_id },
        options: { isList: false },
    }).config({ name: "refetch-cart" });
    refresh_cart_shipping_methods_1.refreshCartShippingMethodsWorkflow.runAsStep({
        input: {
            cart: refetchedCart, // Pass cart to avoid refetch
            additional_data: input.additional_data,
        },
    });
    (0, workflows_sdk_1.when)("force-refresh-update-tax-lines", { input }, ({ input }) => {
        return !!input.force_refresh;
    }).then(() => {
        update_tax_lines_1.updateTaxLinesWorkflow.runAsStep({
            input: { cart_id: input.cart_id },
        });
    });
    (0, workflows_sdk_1.when)("force-refresh-upsert-tax-lines", { input }, ({ input }) => {
        return (!input.force_refresh &&
            (!!input.items?.length || !!input.shipping_methods?.length));
    }).then(() => {
        upsert_tax_lines_1.upsertTaxLinesWorkflow.runAsStep({
            input: (0, workflows_sdk_1.transform)({ refetchedCart, input }, ({ refetchedCart, input }) => {
                return {
                    cart: refetchedCart,
                    items: input.items ?? [],
                    shipping_methods: input.shipping_methods ?? [],
                    force_tax_calculation: input.force_tax_calculation,
                };
            }),
        });
    });
    const cartPromoCodes = (0, workflows_sdk_1.transform)({ refetchedCart, input }, ({ refetchedCart, input }) => {
        if ((0, utils_1.isDefined)(input.promo_codes)) {
            return input.promo_codes;
        }
        else {
            return refetchedCart.promotions.map((p) => p?.code).filter(Boolean);
        }
    });
    update_cart_promotions_1.updateCartPromotionsWorkflow.runAsStep({
        input: {
            cart_id: input.cart_id,
            cart: refetchedCart, // Pass cart to avoid refetch in updateCartPromotionsWorkflow
            promo_codes: cartPromoCodes,
            action: utils_1.PromotionActions.REPLACE,
            force_refresh_payment_collection: false,
        },
    });
    (0, workflows_sdk_1.when)("should-update-item-translations", { input }, ({ input }) => {
        return !!input.locale;
    }).then(() => {
        (0, steps_1.updateCartItemsTranslationsStep)({
            cart_id: input.cart_id,
            locale: input.locale,
            items: refetchedCart.items,
        });
    });
    const beforeRefreshingPaymentCollection = (0, workflows_sdk_1.createHook)("beforeRefreshingPaymentCollection", { input });
    refresh_payment_collection_1.refreshPaymentCollectionForCartWorkflow.runAsStep({
        input: { cart: refetchedCart },
    });
    (0, locking_1.releaseLockStep)({
        key: input.cart_id,
    });
    return new workflows_sdk_1.WorkflowResponse(refetchedCart, {
        hooks: [setPricingContext, beforeRefreshingPaymentCollection],
    });
});
