// @medusajs/link-modules/dist/definitions/user-rbac-role.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UserRbacRole = void 0;
const utils_1 = require("@medusajs/framework/utils");
exports.UserRbacRole = {
    serviceName: utils_1.LINKS.UserRbacRole,
    isLink: true,
    databaseConfig: {
        tableName: "user_rbac_role",
        idPrefix: "userrole",
    },
    alias: [
        {
            name: "user_rbac_role",
        },
        {
            name: "user_rbac_roles",
        },
    ],
    primaryKeys: ["id", "user_id", "rbac_role_id"],
    relationships: [
        {
            serviceName: utils_1.Modules.USER,
            entity: "User",
            primaryKey: "id",
            foreignKey: "user_id",
            alias: "user",
            args: {
                methodSuffix: "Users",
            },
            hasMany: true,
        },
        {
            serviceName: utils_1.Modules.RBAC,
            entity: "RbacRole",
            primaryKey: "id",
            foreignKey: "rbac_role_id",
            alias: "rbac_role",
            args: {
                methodSuffix: "RbacRoles",
            },
            hasMany: true,
        },
    ],
    extends: [
        {
            serviceName: utils_1.Modules.USER,
            entity: "User",
            fieldAlias: {
                rbac_roles: {
                    path: "rbac_roles_link.rbac_role",
                    isList: true,
                },
            },
            relationship: {
                serviceName: utils_1.LINKS.UserRbacRole,
                primaryKey: "user_id",
                foreignKey: "id",
                alias: "rbac_roles_link",
                isList: true,
            },
        },
        {
            serviceName: utils_1.Modules.RBAC,
            entity: "RbacRole",
            relationship: {
                serviceName: utils_1.LINKS.UserRbacRole,
                primaryKey: "rbac_role_id",
                foreignKey: "id",
                alias: "users_link",
                isList: true,
            },
        },
    ],
};
