// @medusajs/medusa/dist/api/admin/rbac/policies/validators.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AdminUpdateRbacPolicy = exports.AdminCreateRbacPolicy = exports.AdminGetRbacPoliciesParams = exports.AdminGetRbacPoliciesParamsFields = exports.AdminGetRbacPolicyParams = void 0;
const zod_1 = require("zod");
const common_validators_1 = require("../../../utils/common-validators");
const validators_1 = require("../../../utils/validators");
exports.AdminGetRbacPolicyParams = (0, validators_1.createSelectParams)();
exports.AdminGetRbacPoliciesParamsFields = zod_1.z.object({
    q: zod_1.z.string().optional(),
    id: zod_1.z.union([zod_1.z.string(), zod_1.z.array(zod_1.z.string())]).optional(),
    key: zod_1.z.union([zod_1.z.string(), zod_1.z.array(zod_1.z.string())]).optional(),
    resource: zod_1.z.union([zod_1.z.string(), zod_1.z.array(zod_1.z.string())]).optional(),
    operation: zod_1.z.union([zod_1.z.string(), zod_1.z.array(zod_1.z.string())]).optional(),
    created_at: (0, validators_1.createOperatorMap)().optional(),
    updated_at: (0, validators_1.createOperatorMap)().optional(),
    deleted_at: (0, validators_1.createOperatorMap)().optional(),
});
exports.AdminGetRbacPoliciesParams = (0, validators_1.createFindParams)({
    limit: 50,
    offset: 0,
})
    .merge(exports.AdminGetRbacPoliciesParamsFields)
    .merge((0, common_validators_1.applyAndAndOrOperators)(exports.AdminGetRbacPoliciesParamsFields));
exports.AdminCreateRbacPolicy = zod_1.z
    .object({
    key: zod_1.z.string(),
    resource: zod_1.z.string(),
    operation: zod_1.z.string(),
    name: zod_1.z.string().nullish(),
    description: zod_1.z.string().nullish(),
    metadata: zod_1.z.record(zod_1.z.unknown()).nullish(),
})
    .strict();
exports.AdminUpdateRbacPolicy = zod_1.z
    .object({
    key: zod_1.z.string().optional(),
    resource: zod_1.z.string().optional(),
    operation: zod_1.z.string().optional(),
    name: zod_1.z.string().nullish(),
    description: zod_1.z.string().nullish(),
    metadata: zod_1.z.record(zod_1.z.unknown()).nullish(),
})
    .strict();
