// @medusajs/core-flows/dist/order/workflows/on-carry-promotions-flag-set.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.onCarryPromotionsFlagSet = exports.onCarryPromotionsFlagSetId = exports.validateCarryPromotionsFlagStep = void 0;
const utils_1 = require("@medusajs/framework/utils");
const workflows_sdk_1 = require("@medusajs/framework/workflows-sdk");
const common_1 = require("../../common");
const order_validation_1 = require("../utils/order-validation");
const compute_adjustments_for_preview_1 = require("./compute-adjustments-for-preview");
const fields_1 = require("./order-edit/utils/fields");
/**
 * This step validates that the order change is an exchange and validates that
 * the promotion allocation is valid for carrying over promotions.
 *
 * :::note
 *
 * You can retrieve details of the order and order change using [Query](https://docs.medusajs.com/learn/fundamentals/module-links/query),
 * or [useQueryGraphStep](https://docs.medusajs.com/resources/references/medusa-workflows/steps/useQueryGraphStep).
 *
 * :::
 *
 * @since 2.12.0
 *
 * @example
 * const data = validateCarryPromotionsFlagStep({
 *   order: {
 *     id: "order_123",
 *     // other order details...
 *   },
 *   orderChange: {
 *     id: "orch_123",
 *     // other order change details...
 *   },
 *   input: {
 *     order_change_id: "orch_123",
 *     carry_over_promotions: true,
 *   }
 * })
 */
exports.validateCarryPromotionsFlagStep = (0, workflows_sdk_1.createStep)("validate-carry-promotions-flag", async function ({ orderChange, order, input, }) {
    // Validate order change is active
    (0, order_validation_1.throwIfOrderChangeIsNotActive)({ orderChange });
    // we don't need to validate promotion since we will be resetting the adjustments
    if (!input.carry_over_promotions) {
        return;
    }
    // Validate promotion allocation if promotions exist
    if (order.promotions && order.promotions.length > 0) {
        const invalidPromotions = [];
        for (const promotion of order.promotions) {
            const applicationMethod = promotion.application_method;
            if (!applicationMethod) {
                continue;
            }
            const allocation = applicationMethod.allocation;
            const type = applicationMethod.type;
            if (allocation !== utils_1.ApplicationMethodAllocation.ACROSS &&
                allocation !== utils_1.ApplicationMethodAllocation.EACH) {
                throw new utils_1.MedusaError(utils_1.MedusaError.Types.INVALID_DATA, `Promotion ${promotion.code || promotion.id} has invalid allocation. Only promotions with EACH or ACROSS allocation can be carried over to outbound exchange items.`);
            }
            // For fixed promotions, allocation must be EACH
            if (type === "fixed" &&
                allocation !== utils_1.ApplicationMethodAllocation.EACH) {
                invalidPromotions.push(promotion.code || promotion.id);
            }
            // For percentage promotions, allocation must be EACH or ACROSS
            if (type === "percentage" &&
                allocation !== utils_1.ApplicationMethodAllocation.EACH &&
                allocation !== utils_1.ApplicationMethodAllocation.ACROSS) {
                invalidPromotions.push(promotion.code || promotion.id);
            }
        }
        if (invalidPromotions.length > 0) {
            throw new utils_1.MedusaError(utils_1.MedusaError.Types.INVALID_DATA, `Promotions with codes ${invalidPromotions.join(", ")} have invalid allocation. Fixed promotions must have EACH allocation, and percentage promotions must have EACH or ACROSS allocation.`);
        }
    }
});
exports.onCarryPromotionsFlagSetId = "on-carry-promotions-flag-set";
/**
 * This workflow toggles whether promotions are carried over to outbound items of an exchange.
 * It validates that the order change is an exchange and that it's active. It also validates that the promotion allocation
 * is valid for carrying over promotions. Finally, it computes adjustments for the order change
 * and either applies or removes promotion adjustments based on whether promotions are to be carried over.
 *
 * This workflow is used by other workflows, such as the {@link updateOrderChangeWorkflow}.
 *
 * You can use this workflow within your customizations or your own custom workflows, allowing you to
 * set the carry over promotions flag for an order change in your custom flows.
 *
 * @since 2.12.0
 *
 * @example
 * const { result } = await onCarryPromotionsFlagSet(container)
 * .run({
 *   input: {
 *     order_change_id: "orch_123",
 *     carry_over_promotions: true,
 *   }
 * })
 *
 * @summary
 *
 * Toggle carrying over promotions to outbound exchange items.
 */
exports.onCarryPromotionsFlagSet = (0, workflows_sdk_1.createWorkflow)(exports.onCarryPromotionsFlagSetId, function (input) {
    const orderChange = (0, common_1.useRemoteQueryStep)({
        entry_point: "order_change",
        fields: [
            "id",
            "status",
            "version",
            "exchange_id",
            "claim_id",
            "return_id",
            "order_id",
            "canceled_at",
            "confirmed_at",
            "declined_at",
            "carry_over_promotions",
        ],
        variables: {
            filters: {
                id: input.order_change_id,
            },
        },
        list: false,
        throw_if_key_not_found: true,
    }).config({ name: "order-change-query" });
    const order = (0, common_1.useRemoteQueryStep)({
        entry_point: "orders",
        fields: [
            ...fields_1.fieldsToComputeAdjustmentsForPreview,
            "promotions.application_method.*",
        ],
        variables: {
            id: orderChange.order_id,
        },
        list: false,
        throw_if_key_not_found: true,
    }).config({ name: "order-query" });
    (0, exports.validateCarryPromotionsFlagStep)({
        orderChange,
        order,
        input,
    });
    const orderWithPromotions = (0, workflows_sdk_1.transform)({ order }, ({ order }) => {
        return {
            ...order,
            promotions: order.promotions ?? [],
        };
    });
    compute_adjustments_for_preview_1.computeAdjustmentsForPreviewWorkflow.runAsStep({
        input: {
            orderChange,
            order: orderWithPromotions,
        },
    });
    return new workflows_sdk_1.WorkflowResponse(void 0);
});
