// @medusajs/utils/dist/modules-sdk/define-policies.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Policy = exports.PolicyOperation = exports.PolicyResource = exports.MedusaPolicySymbol = void 0;
exports.definePolicies = definePolicies;
const common_1 = require("../common");
const to_snake_case_1 = require("../common/to-snake-case");
exports.MedusaPolicySymbol = Symbol.for("MedusaPolicy");
/**
 * Global registry for all unique resources.
 */
exports.PolicyResource = global.PolicyResource ?? {};
global.PolicyResource ??= exports.PolicyResource;
/**
 * Global registry for all unique operations.
 */
const defaultOperations = ["read", "create", "update", "delete", "*"];
exports.PolicyOperation = global.PolicyOperation ?? { ALL: "*" };
global.PolicyOperation ??= exports.PolicyOperation;
for (const operation of defaultOperations) {
    const operationKey = operation === "*" ? "*" : (0, to_snake_case_1.toSnakeCase)(operation);
    exports.PolicyOperation[operationKey] = operation;
}
exports.Policy = global.Policy ?? {};
global.Policy ??= exports.Policy;
/**
 * Define RBAC policies that will be automatically synced to the database
 * when the application starts.
 *
 * @param policies - Single policy or array of policy definitions
 *
 * @example
 * ```ts
 * definePolicies({
 *   name: "ReadBrands",
 *   resource: "brand",
 *   operation: "read"
 *   description: "Read brands"
 * })
 *
 * definePolicies([
 *   {
 *     name: "ReadBrands",
 *     resource: "brand",
 *     operation: "read"
 *   },
 *   {
 *     name: "CreateBrands",
 *     resource: "brand",
 *     operation: "create"
 *   }
 * ])
 * ```
 */
function definePolicies(policies) {
    const callerFilePath = (0, common_1.getCallerFilePath)();
    if ((0, common_1.isFileDisabled)(callerFilePath ?? "")) {
        return { [common_1.MEDUSA_SKIP_FILE]: true };
    }
    const policiesArray = Array.isArray(policies) ? policies : [policies];
    for (const policy of policiesArray) {
        if (!policy.name || !policy.resource || !policy.operation) {
            throw new Error(`Policy definition must include name, resource, and operation. Received: ${JSON.stringify(policy, null, 2)}`);
        }
    }
    for (const policy of policiesArray) {
        const resourceKey = (0, to_snake_case_1.toSnakeCase)(policy.resource);
        const operationKey = (0, to_snake_case_1.toSnakeCase)(policy.operation);
        policy.resource = resourceKey;
        policy.operation = operationKey;
        exports.PolicyResource[resourceKey] = policy.resource;
        exports.PolicyOperation[operationKey] = policy.operation;
        // Register in Policy object with name as key
        exports.Policy[policy.name] = { ...policy };
    }
    const output = {
        [exports.MedusaPolicySymbol]: true,
        policies: policiesArray,
    };
    return output;
}
