// @medusajs/medusa/dist/api/admin/rbac/policies/[id]/route.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DELETE = exports.POST = exports.GET = void 0;
const core_flows_1 = require("@medusajs/core-flows");
const utils_1 = require("@medusajs/framework/utils");
const GET = async (req, res) => {
    const query = req.scope.resolve(utils_1.ContainerRegistrationKeys.QUERY);
    const { data: policies } = await query.graph({
        entity: "rbac_policy",
        filters: { id: req.params.id },
        fields: req.queryConfig.fields,
    });
    const policy = policies[0];
    if (!policy) {
        throw new utils_1.MedusaError(utils_1.MedusaError.Types.NOT_FOUND, `Policy with id: ${req.params.id} not found`);
    }
    res.status(200).json({ policy });
};
exports.GET = GET;
const POST = async (req, res) => {
    const query = req.scope.resolve(utils_1.ContainerRegistrationKeys.QUERY);
    const { data: existing } = await query.graph({
        entity: "rbac_policy",
        filters: { id: req.params.id },
        fields: ["id"],
    });
    const existingPolicy = existing[0];
    if (!existingPolicy) {
        throw new utils_1.MedusaError(utils_1.MedusaError.Types.NOT_FOUND, `Policy with id "${req.params.id}" not found`);
    }
    const { result } = await (0, core_flows_1.updateRbacPoliciesWorkflow)(req.scope).run({
        input: {
            selector: { id: req.params.id },
            update: req.validatedBody,
        },
    });
    const { data: policies } = await query.graph({
        entity: "rbac_policy",
        filters: { id: result[0].id },
        fields: req.queryConfig.fields,
    });
    const policy = policies[0];
    res.status(200).json({ policy });
};
exports.POST = POST;
const DELETE = async (req, res) => {
    const id = req.params.id;
    await (0, core_flows_1.deleteRbacPoliciesWorkflow)(req.scope).run({
        input: { ids: [id] },
    });
    res.status(200).json({
        id,
        object: "rbac_policy",
        deleted: true,
    });
};
exports.DELETE = DELETE;
