// @medusajs/core-flows/dist/cart/steps/prepare-adjustments-from-promotion-actions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.prepareAdjustmentsFromPromotionActionsStep = exports.prepareAdjustmentsFromPromotionActionsStepId = void 0;
const utils_1 = require("@medusajs/framework/utils");
const workflows_sdk_1 = require("@medusajs/framework/workflows-sdk");
exports.prepareAdjustmentsFromPromotionActionsStepId = "prepare-adjustments-from-promotion-actions";
/**
 * This step prepares the line item or shipping method adjustments using
 * actions computed by the Promotion Module.
 *
 * @example
 * const data = prepareAdjustmentsFromPromotionActionsStep({
 *   "actions": [{
 *     "action": "addItemAdjustment",
 *     "item_id": "litem_123",
 *     "amount": 10,
 *     "code": "10OFF",
 *   }]
 * })
 */
exports.prepareAdjustmentsFromPromotionActionsStep = (0, workflows_sdk_1.createStep)(exports.prepareAdjustmentsFromPromotionActionsStepId, async (data, { container }) => {
    const query = container.resolve(utils_1.ContainerRegistrationKeys.QUERY);
    const { actions = [] } = data;
    if (!actions.length) {
        return new workflows_sdk_1.StepResponse({
            lineItemAdjustmentsToCreate: [],
            lineItemAdjustmentIdsToRemove: [],
            shippingMethodAdjustmentsToCreate: [],
            shippingMethodAdjustmentIdsToRemove: [],
            computedPromotionCodes: [],
        });
    }
    const { data: promotions } = await query.graph({
        entity: "promotion",
        fields: ["id", "code"],
        filters: { code: actions.map((a) => a.code) },
    }, { cache: { enable: true } });
    const promotionsMap = new Map(promotions.map((promotion) => [promotion.code, promotion]));
    const lineItemAdjustmentsToCreate = [];
    const lineItemAdjustmentIdsToRemove = [];
    const shippingMethodAdjustmentsToCreate = [];
    const shippingMethodAdjustmentIdsToRemove = [];
    for (const action of actions) {
        switch (action.action) {
            case utils_1.ComputedActions.ADD_ITEM_ADJUSTMENT:
                const itemAction = action;
                lineItemAdjustmentsToCreate.push({
                    code: action.code,
                    amount: itemAction.amount,
                    is_tax_inclusive: itemAction.is_tax_inclusive, // TODO: there is a discrepeancy between the type and the actual data
                    item_id: itemAction.item_id,
                    promotion_id: promotionsMap.get(action.code)?.id,
                });
                break;
            case utils_1.ComputedActions.REMOVE_ITEM_ADJUSTMENT:
                lineItemAdjustmentIdsToRemove.push(action.adjustment_id);
                break;
            case utils_1.ComputedActions.ADD_SHIPPING_METHOD_ADJUSTMENT:
                const shippingAction = action;
                shippingMethodAdjustmentsToCreate.push({
                    code: action.code,
                    amount: shippingAction.amount,
                    shipping_method_id: shippingAction.shipping_method_id,
                    promotion_id: promotionsMap.get(action.code)?.id,
                });
                break;
            case utils_1.ComputedActions.REMOVE_SHIPPING_METHOD_ADJUSTMENT:
                shippingMethodAdjustmentIdsToRemove.push(action.adjustment_id);
                break;
        }
    }
    const computedPromotionCodes = [
        ...lineItemAdjustmentsToCreate,
        ...shippingMethodAdjustmentsToCreate,
    ].map((adjustment) => adjustment.code);
    return new workflows_sdk_1.StepResponse({
        lineItemAdjustmentsToCreate,
        lineItemAdjustmentIdsToRemove,
        shippingMethodAdjustmentsToCreate,
        shippingMethodAdjustmentIdsToRemove,
        computedPromotionCodes,
    });
});
