// @medusajs/core-flows/dist/cart/steps/update-carts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.updateCartsStep = exports.updateCartsStepId = void 0;
const utils_1 = require("@medusajs/framework/utils");
const workflows_sdk_1 = require("@medusajs/framework/workflows-sdk");
exports.updateCartsStepId = "update-carts";
/**
 * This step updates a cart.
 *
 * @example
 * const data = updateCartsStep([{
 *   id: "cart_123",
 *   email: "customer@gmail.com",
 * }])
 */
exports.updateCartsStep = (0, workflows_sdk_1.createStep)(exports.updateCartsStepId, async (data, { container }) => {
    const cartModule = container.resolve(utils_1.Modules.CART);
    if (!data.length) {
        return new workflows_sdk_1.StepResponse([], {
            cartsBeforeUpdate: [],
            addressesBeforeUpdate: [],
        });
    }
    const { selects, relations } = (0, utils_1.getSelectsAndRelationsFromObjectArray)(data, {
        requiredFields: [
            "id",
            "region_id",
            "customer_id",
            "sales_channel_id",
            "email",
            "currency_code",
            "metadata",
            "completed_at",
        ],
    });
    const cartsBeforeUpdate = await cartModule.listCarts({ id: data.map((d) => d.id) }, { select: selects, relations });
    // Since service factory udpate method will correctly keep the reference to the addresses,
    // but won't update its fields, we do this separately
    const addressesInput = data
        .flatMap((cart) => [cart.shipping_address, cart.billing_address])
        .filter((address) => !!address);
    let addressesToUpdateIds = [];
    const addressesToUpdate = addressesInput.filter((address) => {
        if ("id" in address && !!address.id) {
            addressesToUpdateIds.push(address.id);
            return true;
        }
        return false;
    });
    const addressesBeforeUpdate = await cartModule.listAddresses({
        id: addressesToUpdate.map((address) => address.id),
    });
    if (addressesToUpdate.length) {
        await cartModule.updateAddresses(addressesToUpdate);
    }
    const updatedCart = await cartModule.updateCarts(data);
    return new workflows_sdk_1.StepResponse(updatedCart, {
        cartsBeforeUpdate,
        addressesBeforeUpdate,
    });
}, async (dataToCompensate, { container }) => {
    if (!dataToCompensate) {
        return;
    }
    const { cartsBeforeUpdate, addressesBeforeUpdate } = dataToCompensate;
    const cartModule = container.resolve(utils_1.Modules.CART);
    const addressesToUpdate = [];
    for (const address of addressesBeforeUpdate) {
        addressesToUpdate.push({
            ...address,
            metadata: address.metadata ?? undefined,
        });
    }
    await cartModule.updateAddresses(addressesToUpdate);
    const dataToUpdate = [];
    for (const cart of cartsBeforeUpdate) {
        dataToUpdate.push({
            id: cart.id,
            region_id: cart.region_id,
            customer_id: cart.customer_id,
            sales_channel_id: cart.sales_channel_id,
            email: cart.email,
            currency_code: cart.currency_code,
            metadata: cart.metadata,
            completed_at: cart.completed_at,
        });
    }
    return await cartModule.updateCarts(dataToUpdate);
});
