// @medusajs/dashboard/dist/chunk-YIOBBZUB.mjs
import {
  CreateCampaignFormFields
} from "./chunk-5LGRZSEH.mjs";
import {
  useComboboxData
} from "./chunk-3TZOFKX2.mjs";
import {
  Combobox
} from "./chunk-OFN7DIZA.mjs";
import {
  KeyboundForm
} from "./chunk-6HTZNHPT.mjs";
import {
  RouteDrawer
} from "./chunk-WVA4O7QS.mjs";
import {
  useRouteModal
} from "./chunk-D6UW7URG.mjs";
import {
  useDocumentDirection
} from "./chunk-S4DMV3ZT.mjs";
import {
  Form
} from "./chunk-OBQI23QM.mjs";
import {
  useCampaign,
  useUpdatePromotion
} from "./chunk-A63RZVX6.mjs";
import {
  sdk
} from "./chunk-NFEK63OE.mjs";

// src/routes/promotions/promotion-add-campaign/components/add-campaign-promotion-form/add-campaign-promotion-form.tsx
import { zodResolver } from "@hookform/resolvers/zod";
import { Button, RadioGroup, toast } from "@medusajs/ui";
import { useEffect } from "react";
import { useForm, useWatch } from "react-hook-form";
import { useTranslation as useTranslation2 } from "react-i18next";
import * as zod from "zod";

// src/routes/promotions/promotion-add-campaign/components/add-campaign-promotion-form/campaign-details.tsx
import { Heading, Text } from "@medusajs/ui";
import { Fragment } from "react";
import { useTranslation } from "react-i18next";
import { jsx, jsxs } from "react/jsx-runtime";
var translationKeyMap = {
  spend: "spend",
  usage: "usage",
  use_by_attribute: "useByAttribute"
};
var getTranslationKey = (budget) => {
  const translationKey = translationKeyMap[budget?.type] || "-";
  if (budget?.type === "use_by_attribute") {
    if (budget?.attribute === "customer_id") {
      return `campaigns.budget.type.useByAttribute.titleCustomerId`;
    } else if (budget?.attribute === "customer_email") {
      return `campaigns.budget.type.useByAttribute.titleEmail`;
    }
    return `campaigns.budget.type.useByAttribute.title`;
  }
  return `campaigns.budget.type.${translationKey}.title`;
};
var CampaignDetails = ({ campaign }) => {
  const { t } = useTranslation();
  if (!campaign) {
    return;
  }
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    /* @__PURE__ */ jsxs("div", { children: [
      /* @__PURE__ */ jsx(Heading, { level: "h2", className: "mb-4", children: t("campaigns.details") }),
      /* @__PURE__ */ jsxs("div", { className: "text-ui-fg-subtle grid grid-cols-2 items-center py-1", children: [
        /* @__PURE__ */ jsx(Text, { className: "txt-small-plus font-", children: t("campaigns.fields.identifier") }),
        /* @__PURE__ */ jsx("div", { className: "flex items-center gap-1", children: /* @__PURE__ */ jsx(Text, { className: "txt-small", children: campaign.campaign_identifier || "-" }) })
      ] }),
      /* @__PURE__ */ jsxs("div", { className: "text-ui-fg-subtle grid grid-cols-2 items-center py-1", children: [
        /* @__PURE__ */ jsx(Text, { className: "txt-small-plus", children: t("fields.description") }),
        /* @__PURE__ */ jsx("div", { className: "flex items-center gap-1", children: /* @__PURE__ */ jsx(Text, { className: "txt-small", children: campaign.description || "-" }) })
      ] }),
      /* @__PURE__ */ jsxs("div", { className: "text-ui-fg-subtle grid grid-cols-2 items-center py-1", children: [
        /* @__PURE__ */ jsx(Text, { className: "txt-small-plus", children: t("campaigns.fields.start_date") }),
        /* @__PURE__ */ jsx("div", { className: "flex items-center gap-1", children: /* @__PURE__ */ jsx(Text, { className: "txt-small", children: campaign.starts_at?.toString() || "-" }) })
      ] }),
      /* @__PURE__ */ jsxs("div", { className: "text-ui-fg-subtle grid grid-cols-2 items-center py-1", children: [
        /* @__PURE__ */ jsx(Text, { className: "txt-small-plus", children: t("campaigns.fields.end_date") }),
        /* @__PURE__ */ jsx("div", { className: "flex items-center gap-1", children: /* @__PURE__ */ jsx(Text, { className: "txt-small", children: campaign.ends_at?.toString() || "-" }) })
      ] })
    ] }),
    /* @__PURE__ */ jsxs("div", { children: [
      /* @__PURE__ */ jsx(Heading, { level: "h2", className: "mb-4", children: t("campaigns.budget.details") }),
      /* @__PURE__ */ jsxs("div", { className: "text-ui-fg-subtle grid grid-cols-2 items-center py-1", children: [
        /* @__PURE__ */ jsx(Text, { className: "txt-small-plus font-", children: t("campaigns.budget.fields.type") }),
        /* @__PURE__ */ jsx("div", { className: "flex items-center gap-1", children: /* @__PURE__ */ jsx(Text, { className: "txt-small truncate", children: t(getTranslationKey(campaign.budget), {
          defaultValue: "-"
        }) }) })
      ] }),
      /* @__PURE__ */ jsxs("div", { className: "text-ui-fg-subtle grid grid-cols-2 items-center py-1", children: [
        /* @__PURE__ */ jsx(Text, { className: "txt-small-plus", children: t("campaigns.budget.fields.currency") }),
        /* @__PURE__ */ jsx("div", { className: "flex items-center gap-1", children: /* @__PURE__ */ jsx(Text, { className: "txt-small", children: campaign?.budget?.currency_code || "-" }) })
      ] }),
      /* @__PURE__ */ jsxs("div", { className: "text-ui-fg-subtle grid grid-cols-2 items-center py-1", children: [
        /* @__PURE__ */ jsx(Text, { className: "txt-small-plus", children: t("campaigns.budget.fields.limit") }),
        /* @__PURE__ */ jsx("div", { className: "flex items-center gap-1", children: /* @__PURE__ */ jsx(Text, { className: "txt-small", children: campaign.budget?.limit || "-" }) })
      ] }),
      /* @__PURE__ */ jsxs("div", { className: "text-ui-fg-subtle grid grid-cols-2 items-center py-1", children: [
        /* @__PURE__ */ jsx(Text, { className: "txt-small-plus", children: campaign.budget?.type === "use_by_attribute" ? t("campaigns.fields.totalUsedByAttribute") : t("campaigns.budget.fields.used") }),
        /* @__PURE__ */ jsx("div", { className: "flex items-center gap-1", children: /* @__PURE__ */ jsx(Text, { className: "txt-small", children: campaign.budget?.used || "-" }) })
      ] })
    ] })
  ] });
};

// src/routes/promotions/promotion-add-campaign/components/add-campaign-promotion-form/add-campaign-promotion-form.tsx
import { jsx as jsx2, jsxs as jsxs2 } from "react/jsx-runtime";
var EditPromotionSchema = zod.object({
  campaign_id: zod.string().optional().nullable(),
  campaign_choice: zod.enum(["none", "existing"]).optional()
});
var AddCampaignPromotionFields = ({
  form,
  withNewCampaign = true,
  promotionCurrencyCode
}) => {
  const { t } = useTranslation2();
  const direction = useDocumentDirection();
  const watchCampaignId = useWatch({
    control: form.control,
    name: "campaign_id"
  });
  const watchCampaignChoice = useWatch({
    control: form.control,
    name: "campaign_choice"
  });
  const campaignsCombobox = useComboboxData({
    queryFn: (params) => sdk.admin.campaign.list({
      ...params
    }),
    queryKey: ["campaigns"],
    getOptions: (data) => data.campaigns.map((campaign) => ({
      label: campaign.name.toUpperCase(),
      value: campaign.id,
      disabled: campaign.budget?.currency_code && campaign.budget?.currency_code?.toLowerCase() !== promotionCurrencyCode?.toLowerCase()
      // also cannot add promotion which doesn't have currency defined to a campaign with a currency amount budget
    }))
  });
  const { campaign: selectedCampaign } = useCampaign(
    watchCampaignId,
    void 0,
    {
      enabled: !!watchCampaignId
    }
  );
  return /* @__PURE__ */ jsxs2("div", { className: "flex flex-col gap-y-8", children: [
    /* @__PURE__ */ jsx2(
      Form.Field,
      {
        control: form.control,
        name: "campaign_choice",
        render: ({ field }) => {
          return /* @__PURE__ */ jsxs2(Form.Item, { children: [
            /* @__PURE__ */ jsx2(Form.Label, { children: t("promotions.fields.campaign") }),
            /* @__PURE__ */ jsx2(Form.Control, { children: /* @__PURE__ */ jsxs2(
              RadioGroup,
              {
                dir: direction,
                className: "grid grid-cols-1 gap-3",
                ...field,
                value: field.value,
                onValueChange: field.onChange,
                children: [
                  /* @__PURE__ */ jsx2(
                    RadioGroup.ChoiceBox,
                    {
                      value: "none",
                      label: t("promotions.form.campaign.none.title"),
                      description: t("promotions.form.campaign.none.description")
                    }
                  ),
                  /* @__PURE__ */ jsx2(
                    RadioGroup.ChoiceBox,
                    {
                      value: "existing",
                      label: t("promotions.form.campaign.existing.title"),
                      description: t(
                        "promotions.form.campaign.existing.description"
                      )
                    }
                  ),
                  withNewCampaign && /* @__PURE__ */ jsx2(
                    RadioGroup.ChoiceBox,
                    {
                      value: "new",
                      label: t("promotions.form.campaign.new.title"),
                      description: t(
                        "promotions.form.campaign.new.description"
                      )
                    }
                  )
                ]
              }
            ) }),
            /* @__PURE__ */ jsx2(Form.ErrorMessage, {})
          ] });
        }
      }
    ),
    watchCampaignChoice === "existing" && /* @__PURE__ */ jsx2(
      Form.Field,
      {
        control: form.control,
        name: "campaign_id",
        render: ({ field: { onChange, ...field } }) => {
          return /* @__PURE__ */ jsxs2(Form.Item, { children: [
            /* @__PURE__ */ jsx2(Form.Label, { tooltip: t("campaigns.fields.campaign_id.hint"), children: t("promotions.form.campaign.existing.title") }),
            /* @__PURE__ */ jsx2(Form.Control, { children: /* @__PURE__ */ jsx2(
              Combobox,
              {
                dir: direction,
                options: campaignsCombobox.options,
                searchValue: campaignsCombobox.searchValue,
                onSearchValueChange: campaignsCombobox.onSearchValueChange,
                onChange,
                ...field
              }
            ) }),
            /* @__PURE__ */ jsx2(Form.ErrorMessage, {})
          ] });
        }
      }
    ),
    watchCampaignChoice === "new" && /* @__PURE__ */ jsx2(CreateCampaignFormFields, { form, fieldScope: "campaign." }),
    /* @__PURE__ */ jsx2(CampaignDetails, { campaign: selectedCampaign })
  ] });
};
var AddCampaignPromotionForm = ({
  promotion
}) => {
  const { t } = useTranslation2();
  const { handleSuccess } = useRouteModal();
  const { campaign } = promotion;
  const originalId = campaign?.id;
  const form = useForm({
    defaultValues: {
      campaign_id: campaign?.id,
      campaign_choice: campaign?.id ? "existing" : "none"
    },
    resolver: zodResolver(EditPromotionSchema)
  });
  const { setValue } = form;
  const { mutateAsync, isPending } = useUpdatePromotion(promotion.id);
  const handleSubmit = form.handleSubmit(async (data) => {
    await mutateAsync(
      { campaign_id: data.campaign_id },
      {
        onSuccess: () => {
          toast.success(t("promotions.campaign.edit.successToast"));
          handleSuccess();
        },
        onError: (e) => {
          toast.error(e.message);
        }
      }
    );
  });
  const watchCampaignChoice = useWatch({
    control: form.control,
    name: "campaign_choice"
  });
  useEffect(() => {
    if (watchCampaignChoice === "none") {
      setValue("campaign_id", null);
    }
    if (watchCampaignChoice === "existing") {
      setValue("campaign_id", originalId);
    }
  }, [watchCampaignChoice, setValue, originalId]);
  return /* @__PURE__ */ jsx2(RouteDrawer.Form, { form, children: /* @__PURE__ */ jsxs2(
    KeyboundForm,
    {
      onSubmit: handleSubmit,
      className: "flex size-full flex-col overflow-hidden",
      children: [
        /* @__PURE__ */ jsx2(RouteDrawer.Body, { className: "size-full overflow-auto", children: /* @__PURE__ */ jsx2(
          AddCampaignPromotionFields,
          {
            form,
            withNewCampaign: false,
            promotionCurrencyCode: promotion.application_method?.currency_code
          }
        ) }),
        /* @__PURE__ */ jsx2(RouteDrawer.Footer, { children: /* @__PURE__ */ jsxs2("div", { className: "flex items-center justify-end gap-x-2", children: [
          /* @__PURE__ */ jsx2(RouteDrawer.Close, { asChild: true, children: /* @__PURE__ */ jsx2(Button, { size: "small", variant: "secondary", children: t("actions.cancel") }) }),
          /* @__PURE__ */ jsx2(Button, { size: "small", type: "submit", isLoading: isPending, children: t("actions.save") })
        ] }) })
      ]
    }
  ) });
};

export {
  AddCampaignPromotionFields,
  AddCampaignPromotionForm
};
