// @medusajs/dashboard/dist/promotion-edit-details-6BSOOWQN.mjs
import {
  DeprecatedPercentageInput
} from "./chunk-KEQOAJC2.mjs";
import {
  currencies,
  getCurrencySymbol
} from "./chunk-DK4WIVY6.mjs";
import {
  SwitchBox
} from "./chunk-PRLQLEEQ.mjs";
import "./chunk-IUCDCPJU.mjs";
import {
  KeyboundForm
} from "./chunk-6HTZNHPT.mjs";
import {
  RouteDrawer
} from "./chunk-WVA4O7QS.mjs";
import {
  useRouteModal
} from "./chunk-D6UW7URG.mjs";
import {
  useDocumentDirection
} from "./chunk-S4DMV3ZT.mjs";
import {
  Form
} from "./chunk-OBQI23QM.mjs";
import {
  usePromotion,
  useUpdatePromotion
} from "./chunk-A63RZVX6.mjs";
import "./chunk-FXYH54JP.mjs";
import "./chunk-774WSTCC.mjs";
import "./chunk-NFEK63OE.mjs";
import "./chunk-QZ7TP4HQ.mjs";

// src/routes/promotions/promotion-edit-details/promotion-edit-details.tsx
import { Heading } from "@medusajs/ui";
import { useTranslation as useTranslation2 } from "react-i18next";
import { useParams } from "react-router-dom";

// src/routes/promotions/promotion-edit-details/components/edit-promotion-form/edit-promotion-details-form.tsx
import { zodResolver } from "@hookform/resolvers/zod";
import { Button, CurrencyInput, Input, RadioGroup, Text } from "@medusajs/ui";
import { useForm, useWatch } from "react-hook-form";
import { Trans, useTranslation } from "react-i18next";
import { useEffect } from "react";
import * as zod from "zod";
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
var EditPromotionSchema = zod.object({
  is_automatic: zod.string().toLowerCase(),
  code: zod.string().min(1),
  is_tax_inclusive: zod.boolean().optional(),
  status: zod.enum(["active", "inactive", "draft"]),
  value_type: zod.enum(["fixed", "percentage"]),
  value: zod.number().min(0).or(zod.string().min(1)),
  allocation: zod.enum(["each", "across", "once"]),
  target_type: zod.enum(["order", "shipping_methods", "items"]),
  max_quantity: zod.number().min(1).optional().nullable()
});
var EditPromotionDetailsForm = ({
  promotion
}) => {
  const { t } = useTranslation();
  const { handleSuccess } = useRouteModal();
  const form = useForm({
    defaultValues: {
      is_automatic: promotion.is_automatic.toString(),
      is_tax_inclusive: promotion.is_tax_inclusive,
      code: promotion.code,
      status: promotion.status,
      value: promotion.application_method.value,
      allocation: promotion.application_method.allocation,
      value_type: promotion.application_method.type,
      target_type: promotion.application_method.target_type,
      max_quantity: promotion.application_method.max_quantity
    },
    resolver: zodResolver(EditPromotionSchema)
  });
  const watchValueType = useWatch({
    control: form.control,
    name: "value_type"
  });
  const watchAllocation = useWatch({
    control: form.control,
    name: "allocation"
  });
  const isFixedValueType = watchValueType === "fixed";
  const originalAllocation = promotion.application_method.allocation;
  const { mutateAsync, isPending } = useUpdatePromotion(promotion.id);
  const handleSubmit = form.handleSubmit(async (data) => {
    const value = parseFloat(data.value);
    if (isNaN(value) || value < 0) {
      form.setError("value", { message: t("promotions.form.value.invalid") });
      return;
    }
    await mutateAsync(
      {
        is_automatic: data.is_automatic === "true",
        code: data.code,
        status: data.status,
        is_tax_inclusive: data.is_tax_inclusive,
        application_method: {
          value: parseFloat(data.value),
          type: data.value_type,
          allocation: data.allocation,
          max_quantity: data.max_quantity
        }
      },
      {
        onSuccess: () => {
          handleSuccess();
        }
      }
    );
  });
  const allocationWatchValue = useWatch({
    control: form.control,
    name: "value_type"
  });
  useEffect(() => {
    if (!(allocationWatchValue === "fixed" && promotion.type === "standard")) {
      form.setValue("is_tax_inclusive", false);
    }
  }, [allocationWatchValue, form, promotion]);
  const direction = useDocumentDirection();
  return /* @__PURE__ */ jsx(RouteDrawer.Form, { form, children: /* @__PURE__ */ jsxs(
    KeyboundForm,
    {
      onSubmit: handleSubmit,
      className: "flex flex-1 flex-col overflow-hidden",
      children: [
        /* @__PURE__ */ jsx(RouteDrawer.Body, { className: "flex flex-1 flex-col gap-y-8 overflow-y-auto", children: /* @__PURE__ */ jsxs("div", { className: "flex flex-col gap-y-8", children: [
          /* @__PURE__ */ jsx(
            Form.Field,
            {
              control: form.control,
              name: "status",
              render: ({ field }) => {
                return /* @__PURE__ */ jsxs(Form.Item, { children: [
                  /* @__PURE__ */ jsx(Form.Label, { children: t("promotions.form.status.label") }),
                  /* @__PURE__ */ jsx(Form.Control, { children: /* @__PURE__ */ jsxs(
                    RadioGroup,
                    {
                      dir: direction,
                      className: "flex-col gap-y-3",
                      ...field,
                      value: field.value,
                      onValueChange: field.onChange,
                      children: [
                        /* @__PURE__ */ jsx(
                          RadioGroup.ChoiceBox,
                          {
                            value: "draft",
                            label: t("promotions.form.status.draft.title"),
                            description: t(
                              "promotions.form.status.draft.description"
                            )
                          }
                        ),
                        /* @__PURE__ */ jsx(
                          RadioGroup.ChoiceBox,
                          {
                            value: "active",
                            label: t("promotions.form.status.active.title"),
                            description: t(
                              "promotions.form.status.active.description"
                            )
                          }
                        ),
                        /* @__PURE__ */ jsx(
                          RadioGroup.ChoiceBox,
                          {
                            value: "inactive",
                            label: t("promotions.form.status.inactive.title"),
                            description: t(
                              "promotions.form.status.inactive.description"
                            )
                          }
                        )
                      ]
                    }
                  ) }),
                  /* @__PURE__ */ jsx(Form.ErrorMessage, {})
                ] });
              }
            }
          ),
          /* @__PURE__ */ jsx(
            Form.Field,
            {
              control: form.control,
              name: "is_automatic",
              render: ({ field }) => {
                return /* @__PURE__ */ jsxs(Form.Item, { children: [
                  /* @__PURE__ */ jsx(Form.Label, { children: t("promotions.form.method.label") }),
                  /* @__PURE__ */ jsx(Form.Control, { children: /* @__PURE__ */ jsxs(
                    RadioGroup,
                    {
                      dir: direction,
                      className: "flex-col gap-y-3",
                      ...field,
                      value: field.value,
                      onValueChange: field.onChange,
                      children: [
                        /* @__PURE__ */ jsx(
                          RadioGroup.ChoiceBox,
                          {
                            value: "false",
                            label: t("promotions.form.method.code.title"),
                            description: t(
                              "promotions.form.method.code.description"
                            )
                          }
                        ),
                        /* @__PURE__ */ jsx(
                          RadioGroup.ChoiceBox,
                          {
                            value: "true",
                            label: t("promotions.form.method.automatic.title"),
                            description: t(
                              "promotions.form.method.automatic.description"
                            )
                          }
                        )
                      ]
                    }
                  ) }),
                  /* @__PURE__ */ jsx(Form.ErrorMessage, {})
                ] });
              }
            }
          ),
          allocationWatchValue === "fixed" && promotion.type === "standard" && /* @__PURE__ */ jsx(
            SwitchBox,
            {
              control: form.control,
              name: "is_tax_inclusive",
              label: t("promotions.form.taxInclusive.title"),
              description: t("promotions.form.taxInclusive.description")
            }
          ),
          /* @__PURE__ */ jsxs("div", { className: "flex flex-col gap-y-4", children: [
            /* @__PURE__ */ jsx(
              Form.Field,
              {
                control: form.control,
                name: "code",
                render: ({ field }) => {
                  return /* @__PURE__ */ jsxs(Form.Item, { children: [
                    /* @__PURE__ */ jsx(Form.Label, { children: t("promotions.form.code.title") }),
                    /* @__PURE__ */ jsx(Form.Control, { children: /* @__PURE__ */ jsx(Input, { ...field }) })
                  ] });
                }
              }
            ),
            /* @__PURE__ */ jsx(
              Text,
              {
                size: "small",
                leading: "compact",
                className: "text-ui-fg-subtle",
                children: /* @__PURE__ */ jsx(
                  Trans,
                  {
                    t,
                    i18nKey: "promotions.form.code.description",
                    components: [/* @__PURE__ */ jsx("br", {}, "break")]
                  }
                )
              }
            )
          ] }),
          promotion.application_method?.target_type !== "shipping_methods" && /* @__PURE__ */ jsxs(Fragment, { children: [
            /* @__PURE__ */ jsx(
              Form.Field,
              {
                control: form.control,
                name: "value_type",
                render: ({ field }) => {
                  return /* @__PURE__ */ jsxs(Form.Item, { children: [
                    /* @__PURE__ */ jsx(Form.Label, { children: t("promotions.fields.value_type") }),
                    /* @__PURE__ */ jsx(Form.Control, { children: /* @__PURE__ */ jsxs(
                      RadioGroup,
                      {
                        dir: direction,
                        className: "flex-col gap-y-3",
                        ...field,
                        onValueChange: field.onChange,
                        children: [
                          /* @__PURE__ */ jsx(
                            RadioGroup.ChoiceBox,
                            {
                              value: "fixed",
                              label: t(
                                "promotions.form.value_type.fixed.title"
                              ),
                              description: t(
                                "promotions.form.value_type.fixed.description"
                              )
                            }
                          ),
                          /* @__PURE__ */ jsx(
                            RadioGroup.ChoiceBox,
                            {
                              value: "percentage",
                              label: t(
                                "promotions.form.value_type.percentage.title"
                              ),
                              description: t(
                                "promotions.form.value_type.percentage.description"
                              )
                            }
                          )
                        ]
                      }
                    ) }),
                    /* @__PURE__ */ jsx(Form.ErrorMessage, {})
                  ] });
                }
              }
            ),
            /* @__PURE__ */ jsx(
              Form.Field,
              {
                control: form.control,
                name: "value",
                render: ({ field: { onChange, ...field } }) => {
                  const currencyCode = promotion.application_method?.currency_code ?? "USD";
                  const currencyInfo = currencies[currencyCode?.toUpperCase() || "USD"];
                  return /* @__PURE__ */ jsxs(Form.Item, { children: [
                    /* @__PURE__ */ jsx(Form.Label, { children: isFixedValueType ? t("fields.amount") : t("fields.percentage") }),
                    /* @__PURE__ */ jsx(Form.Control, { children: isFixedValueType ? /* @__PURE__ */ jsx(
                      CurrencyInput,
                      {
                        min: 0,
                        onValueChange: (val) => onChange(val),
                        decimalSeparator: ".",
                        groupSeparator: ",",
                        decimalScale: currencyInfo.decimal_digits,
                        decimalsLimit: currencyInfo.decimal_digits,
                        code: currencyCode,
                        symbol: getCurrencySymbol(currencyCode),
                        ...field,
                        value: field.value
                      }
                    ) : /* @__PURE__ */ jsx(
                      DeprecatedPercentageInput,
                      {
                        min: 0,
                        max: 100,
                        ...field,
                        value: field.value || "",
                        onChange: (e) => {
                          onChange(
                            e.target.value === "" ? null : parseFloat(e.target.value)
                          );
                        }
                      },
                      "amount"
                    ) }),
                    /* @__PURE__ */ jsx(Form.ErrorMessage, {})
                  ] });
                }
              }
            ),
            /* @__PURE__ */ jsx(
              Form.Field,
              {
                control: form.control,
                name: "allocation",
                render: ({ field }) => {
                  return /* @__PURE__ */ jsxs(Form.Item, { children: [
                    /* @__PURE__ */ jsx(
                      Form.Label,
                      {
                        tooltip: t("promotions.fields.allocationTooltip"),
                        children: t("promotions.fields.allocation")
                      }
                    ),
                    /* @__PURE__ */ jsx(Form.Control, { children: /* @__PURE__ */ jsxs(
                      RadioGroup,
                      {
                        dir: direction,
                        className: "flex-col gap-y-3",
                        ...field,
                        onValueChange: field.onChange,
                        children: [
                          /* @__PURE__ */ jsx(
                            RadioGroup.ChoiceBox,
                            {
                              value: "each",
                              label: t("promotions.form.allocation.each.title"),
                              description: t(
                                "promotions.form.allocation.each.description"
                              ),
                              disabled: originalAllocation === "across"
                            }
                          ),
                          /* @__PURE__ */ jsx(
                            RadioGroup.ChoiceBox,
                            {
                              value: "across",
                              label: t(
                                "promotions.form.allocation.across.title"
                              ),
                              description: t(
                                "promotions.form.allocation.across.description"
                              ),
                              disabled: originalAllocation === "each" || originalAllocation === "once"
                            }
                          ),
                          /* @__PURE__ */ jsx(
                            RadioGroup.ChoiceBox,
                            {
                              value: "once",
                              label: t("promotions.form.allocation.once.title"),
                              description: t(
                                "promotions.form.allocation.once.description"
                              ),
                              disabled: originalAllocation === "across"
                            }
                          )
                        ]
                      }
                    ) }),
                    /* @__PURE__ */ jsx(Form.ErrorMessage, {})
                  ] });
                }
              }
            ),
            (watchAllocation === "each" || watchAllocation === "once") && /* @__PURE__ */ jsx(
              Form.Field,
              {
                control: form.control,
                name: "max_quantity",
                render: ({ field }) => {
                  return /* @__PURE__ */ jsxs(Form.Item, { children: [
                    /* @__PURE__ */ jsx(Form.Label, { children: t("promotions.form.max_quantity.title") }),
                    /* @__PURE__ */ jsx(Form.Control, { children: /* @__PURE__ */ jsx(
                      Input,
                      {
                        ...form.register("max_quantity", {
                          valueAsNumber: true
                        }),
                        type: "number",
                        min: 1,
                        placeholder: "3"
                      }
                    ) }),
                    /* @__PURE__ */ jsx(
                      Text,
                      {
                        size: "small",
                        leading: "compact",
                        className: "text-ui-fg-subtle",
                        children: /* @__PURE__ */ jsx(
                          Trans,
                          {
                            t,
                            i18nKey: "promotions.form.max_quantity.description",
                            components: [/* @__PURE__ */ jsx("br", {}, "break")]
                          }
                        )
                      }
                    ),
                    /* @__PURE__ */ jsx(Form.ErrorMessage, {})
                  ] });
                }
              }
            )
          ] })
        ] }) }),
        /* @__PURE__ */ jsx(RouteDrawer.Footer, { children: /* @__PURE__ */ jsxs("div", { className: "flex items-center justify-end gap-x-2", children: [
          /* @__PURE__ */ jsx(RouteDrawer.Close, { asChild: true, children: /* @__PURE__ */ jsx(Button, { size: "small", variant: "secondary", children: t("actions.cancel") }) }),
          /* @__PURE__ */ jsx(Button, { size: "small", type: "submit", isLoading: isPending, children: t("actions.save") })
        ] }) })
      ]
    }
  ) });
};

// src/routes/promotions/promotion-edit-details/promotion-edit-details.tsx
import { jsx as jsx2, jsxs as jsxs2 } from "react/jsx-runtime";
var PromotionEditDetails = () => {
  const { id } = useParams();
  const { t } = useTranslation2();
  const { promotion, isLoading, isError, error } = usePromotion(id);
  if (isError) {
    throw error;
  }
  return /* @__PURE__ */ jsxs2(RouteDrawer, { children: [
    /* @__PURE__ */ jsx2(RouteDrawer.Header, { children: /* @__PURE__ */ jsx2(Heading, { children: t("promotions.edit.title") }) }),
    !isLoading && promotion && /* @__PURE__ */ jsx2(EditPromotionDetailsForm, { promotion })
  ] });
};
export {
  PromotionEditDetails as Component
};
