// @medusajs/core-flows/dist/draft-order/workflows/compute-draft-order-adjustments.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.computeDraftOrderAdjustmentsWorkflow = exports.computeDraftOrderAdjustmentsWorkflowId = void 0;
const utils_1 = require("@medusajs/framework/utils");
const workflows_sdk_1 = require("@medusajs/framework/workflows-sdk");
const cart_1 = require("../../cart");
const create_order_change_actions_1 = require("../../order/workflows/create-order-change-actions");
const preview_order_change_1 = require("../../order/steps/preview-order-change");
const validate_draft_order_change_1 = require("../steps/validate-draft-order-change");
const fields_1 = require("../utils/fields");
const common_1 = require("../../common");
const locking_1 = require("../../locking");
const delete_order_change_actions_1 = require("../../order/steps/delete-order-change-actions");
const prepare_order_compute_action_context_1 = require("../../order/workflows/order-edit/utils/prepare-order-compute-action-context");
exports.computeDraftOrderAdjustmentsWorkflowId = "compute-draft-order-adjustments";
/**
 * This workflow computes the adjustments or promotions for a draft order. It's used by other workflows
 * to compute new adjustments for the promotions whenever changes are made to the draft order.
 * Created adjustments are "virtual" meaning they live on the action and no line item adjustments records are created
 * in the database until the edit is confirmed.
 *
 * You can use this workflow within your customizations or your own custom workflows, allowing you to wrap custom logic around
 * computing the adjustments for a draft order.
 *
 * @example
 * const { result } = await computeDraftOrderAdjustmentsWorkflow(container)
 * .run({
 *   input: {
 *     order_id: "order_123",
 *   }
 * })
 *
 * @summary
 *
 * Refresh the promotions in a draft order.
 */
exports.computeDraftOrderAdjustmentsWorkflow = (0, workflows_sdk_1.createWorkflow)(exports.computeDraftOrderAdjustmentsWorkflowId, function (input) {
    (0, locking_1.acquireLockStep)({
        key: input.order_id,
        timeout: 2,
        ttl: 10,
    });
    const order = (0, common_1.useRemoteQueryStep)({
        entry_point: "orders",
        fields: fields_1.draftOrderFieldsForRefreshSteps,
        variables: { id: input.order_id },
        list: false,
        throw_if_key_not_found: true,
    }).config({ name: "order-query" });
    const orderChange = (0, common_1.useRemoteQueryStep)({
        entry_point: "order_change",
        fields: ["id", "status", "version", "actions.*"],
        variables: {
            filters: {
                order_id: input.order_id,
                status: [utils_1.OrderChangeStatus.PENDING, utils_1.OrderChangeStatus.REQUESTED],
            },
        },
        list: false,
    }).config({ name: "order-change-query" });
    (0, validate_draft_order_change_1.validateDraftOrderChangeStep)({ order, orderChange });
    const toDeleteActions = (0, workflows_sdk_1.transform)(orderChange, (orderChange) => {
        return orderChange.actions
            .filter((action) => action.action === utils_1.ChangeActionType.ITEM_ADJUSTMENTS_REPLACE)
            .map((action) => {
            return action.id;
        });
    });
    (0, workflows_sdk_1.when)(toDeleteActions, (toDeleteActions) => toDeleteActions.length > 0).then(() => {
        // clean up old replace actions from the current order change
        (0, delete_order_change_actions_1.deleteOrderChangeActionsStep)({ ids: toDeleteActions });
    });
    const previewedOrder = (0, preview_order_change_1.previewOrderChangeStep)(input.order_id);
    (0, workflows_sdk_1.when)({ order }, ({ order }) => Array.isArray(order.promotions) && !order.promotions.length).then(() => {
        const orderChangeActionAdjustmentsInput = (0, workflows_sdk_1.transform)({ order, previewedOrder, orderChange }, ({ order, previewedOrder, orderChange }) => {
            return previewedOrder.items.map((item) => {
                return {
                    order_id: order.id,
                    order_change_id: orderChange.id,
                    version: orderChange.version,
                    action: utils_1.ChangeActionType.ITEM_ADJUSTMENTS_REPLACE,
                    details: {
                        reference_id: item.id,
                        adjustments: [],
                    },
                };
            });
        });
        create_order_change_actions_1.createOrderChangeActionsWorkflow
            .runAsStep({ input: orderChangeActionAdjustmentsInput })
            .config({ name: "order-change-action-adjustments-input-remove" });
    });
    (0, workflows_sdk_1.when)({ order }, ({ order }) => !!order.promotions?.length).then(() => {
        const orderPromotions = (0, workflows_sdk_1.transform)({ order }, ({ order }) => {
            return order.promotions
                .map((p) => p.code)
                .filter((p) => p !== undefined);
        });
        const actionsToComputeItemsInput = (0, prepare_order_compute_action_context_1.prepareOrderComputeActionContextStep)({
            order,
            previewedOrder,
        });
        const actions = (0, cart_1.getActionsToComputeFromPromotionsStep)({
            computeActionContext: actionsToComputeItemsInput,
            promotionCodesToApply: orderPromotions,
        });
        const { lineItemAdjustmentsToCreate } = (0, cart_1.prepareAdjustmentsFromPromotionActionsStep)({ actions });
        const orderChangeActionAdjustmentsInput = (0, workflows_sdk_1.transform)({
            order,
            previewedOrder,
            orderChange,
            lineItemAdjustmentsToCreate,
        }, ({ order, previewedOrder, orderChange, lineItemAdjustmentsToCreate, }) => {
            return previewedOrder.items.map((item) => {
                const itemAdjustments = lineItemAdjustmentsToCreate.filter((adjustment) => adjustment.item_id === item.id);
                return {
                    order_change_id: orderChange.id,
                    order_id: order.id,
                    version: orderChange.version,
                    action: utils_1.ChangeActionType.ITEM_ADJUSTMENTS_REPLACE,
                    details: {
                        reference_id: item.id,
                        adjustments: itemAdjustments,
                    },
                };
            });
        });
        create_order_change_actions_1.createOrderChangeActionsWorkflow
            .runAsStep({ input: orderChangeActionAdjustmentsInput })
            .config({ name: "order-change-action-adjustments-input" });
    });
    (0, locking_1.releaseLockStep)({
        key: input.order_id,
    });
    return new workflows_sdk_1.WorkflowResponse(void 0);
});
