// @medusajs/medusa/dist/api/admin/rbac/roles/validators.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AdminAddRolePoliciesType = exports.AdminUpdateRbacRole = exports.AdminCreateRbacRole = exports.AdminGetRbacRolesParams = exports.AdminGetRbacRolesParamsFields = exports.AdminGetRbacRoleParams = void 0;
const zod_1 = require("zod");
const common_validators_1 = require("../../../utils/common-validators");
const validators_1 = require("../../../utils/validators");
exports.AdminGetRbacRoleParams = (0, validators_1.createSelectParams)().merge(zod_1.z.object({
    policies: zod_1.z.union([zod_1.z.string(), zod_1.z.array(zod_1.z.string())]).optional(),
}));
exports.AdminGetRbacRolesParamsFields = zod_1.z.object({
    q: zod_1.z.string().optional(),
    id: zod_1.z.union([zod_1.z.string(), zod_1.z.array(zod_1.z.string())]).optional(),
    name: zod_1.z.union([zod_1.z.string(), zod_1.z.array(zod_1.z.string())]).optional(),
    parent_id: zod_1.z.union([zod_1.z.string(), zod_1.z.array(zod_1.z.string())]).optional(),
    created_at: (0, validators_1.createOperatorMap)().optional(),
    updated_at: (0, validators_1.createOperatorMap)().optional(),
    deleted_at: (0, validators_1.createOperatorMap)().optional(),
});
exports.AdminGetRbacRolesParams = (0, validators_1.createFindParams)({
    limit: 50,
    offset: 0,
})
    .merge(exports.AdminGetRbacRolesParamsFields)
    .merge((0, common_validators_1.applyAndAndOrOperators)(exports.AdminGetRbacRolesParamsFields));
exports.AdminCreateRbacRole = zod_1.z
    .object({
    name: zod_1.z.string(),
    parent_id: zod_1.z.string().nullish(),
    description: zod_1.z.string().nullish(),
    metadata: zod_1.z.record(zod_1.z.unknown()).nullish(),
})
    .strict();
exports.AdminUpdateRbacRole = zod_1.z
    .object({
    name: zod_1.z.string().optional(),
    parent_id: zod_1.z.string().nullish(),
    description: zod_1.z.string().nullish(),
    metadata: zod_1.z.record(zod_1.z.unknown()).nullish(),
})
    .strict();
exports.AdminAddRolePoliciesType = zod_1.z.object({
    policies: zod_1.z.array(zod_1.z.string().min(1)).min(1),
});
