// @medusajs/medusa/dist/commands/utils/generate-types.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateTypes = generateTypes;
const framework_1 = require("@medusajs/framework");
const modules_sdk_1 = require("@medusajs/framework/modules-sdk");
const utils_1 = require("@medusajs/framework/utils");
const path_1 = __importStar(require("path"));
async function generateTypes({ directory, container, logger, }) {
    logger.info("Generating types...");
    const configModule = container.resolve(utils_1.ContainerRegistrationKeys.CONFIG_MODULE);
    const plugins = await (0, utils_1.getResolvedPlugins)(directory, configModule, true);
    (0, utils_1.mergePluginModules)(configModule, plugins);
    Object.keys(configModule.modules ?? {}).forEach((key) => {
        (0, utils_1.validateModuleName)(key);
    });
    const linksSourcePaths = plugins.map((plugin) => (0, path_1.join)(plugin.resolve, "links"));
    await new framework_1.LinkLoader(linksSourcePaths, logger).load();
    const { gqlSchema, modules } = await new framework_1.MedusaAppLoader().load({
        registerInContainer: false,
        schemaOnly: true,
    });
    const typesDirectory = path_1.default.join(directory, ".medusa/types");
    /**
     * Cleanup existing types directory before creating new artifacts
     */
    await new utils_1.FileSystem(typesDirectory).cleanup({ recursive: true });
    await (0, utils_1.generateContainerTypes)(modules, {
        outputDir: typesDirectory,
        interfaceName: "ModuleImplementations",
    });
    logger.debug("Generated container types");
    if (gqlSchema) {
        await (0, utils_1.gqlSchemaToTypes)({
            outputDir: typesDirectory,
            filename: "query-entry-points",
            interfaceName: "RemoteQueryEntryPoints",
            schema: gqlSchema,
            joinerConfigs: modules_sdk_1.MedusaModule.getAllJoinerConfigs(),
        });
        logger.debug("Generated modules types");
    }
    logger.info("Types generated successfully");
}
